File: SatelliteAndroid/satellite/src/main/java/yuni/library/satellite/Satellite.java
package yuni.library.satellite;

import android.content.Context;
import android.util.Log;

import java.util.Objects;

public class Satellite {

    public static final String TAG = Satellite.class.getName();

    /* 精度模式 */
    public static final int MODE_LOW = 100;
    public static final int MODE_FINE = 200;
    public static final int MODE_HIGH = 300;

    public static final Satellite shared = new Satellite();

    /**
     * 定位参数配置
     */
    public static class Options {

        /**
         *  当once为false，此值有效，持续定位间隔
         */
        public long interval = 10000;

        /**
         * 是否返回逆地理位置信息
         */
        public boolean address = false;

        /**
         * 当cacheTime大于0，设置可获取缓存位置时间，如果最后一次位置超时了则重新获取
         */
        public long cacheTime = 0;

        /**
         * 精度模式
         */
        public int mode = MODE_HIGH;

        public boolean isCache() {
            return cacheTime > 0;
        }

        public Options copy() {
            Options options = new Options();
            options.interval = interval;
            options.address = address;
            options.cacheTime = cacheTime;
            options.mode = mode;
            return options;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Options options = (Options) o;
            return interval == options.interval &&
                    address == options.address &&
                    cacheTime == options.cacheTime &&
                    mode == options.mode;
        }

        @Override
        public int hashCode() {
            return Objects.hash(interval, address, cacheTime, mode);
        }

        @Override
        public String toString() {
            return "Options{" +
                    ", interval=" + interval +
                    ", address=" + address +
                    ", cacheTime=" + cacheTime +
                    ", mode=" + mode +
                    '}';
        }
    }

    public interface Logger {
        void log(int level, String tag, String message);
    }

    public interface Listener {
        void onLocationChanged(Location location);
    }

    Logger mLogger = new Logger() {
        @Override
        public void log(int level, String tag,  String message) {
            Log.println(level, tag, message);
        }
    };

    private LocationService mLocationService;
    private LocationManager mLocationManager;


    Satellite() {}

    public void init(Context context) {
       this.init(context, new AMapLocationService(context));
    }

    public void init(Context context, LocationService locationService) {
        mLocationService = locationService;
        mLocationManager = new LocationManager(mLocationService);
    }

    public void log(int level, String message) {
        mLogger.log(level, TAG, message);
    }

    public void setLogger(Logger logger) {
        if (logger == null) {
            return;
        }
        mLogger = logger;
    }

    public void getLocationOnce(Options options, Listener listener) {
        mLocationManager.getLocationOnce(options, listener);
    }

    public LocationHandler createContinue(Options options, Listener listener) {
        return new LocationHandler(options, listener);
    }

    public Location getLastLocation() {
        return mLocationManager.getLastLocation();
    }

    public class LocationHandler {

        private Options mOptions;
        private Listener mListener;
        private LocationManager.ContinueLocationHolder mHolder;

        LocationHandler(Options options, Listener listener) {
            mOptions = options;
            mListener = listener;
        }

        public void setOptions(Options options) {
            mOptions = options;
            if (mHolder != null) {
                mLocationManager.updateOptions(mHolder, mOptions);
            }
        }

        public void stop() {
            if (mHolder == null) {
                return;
            }
            mLocationManager.stopContinue(mHolder);
            mHolder = null;
        }

        public void start() {
            if (mHolder != null) {
                return;
            }
            mHolder = mLocationManager.startContinue(this, mOptions, mListener);
        }

    }

}

File: SatelliteIOS/SatelliteIOS/Classes/Satellite.swift
import Foundation
import CoreLocation

public typealias SatelliteLogFunc = (_ message: String) -> Void
public typealias SatelliteListener = (_ error: Error?, _ location: CLLocation?) -> Void

public protocol SatelliteLogDelegate {
    func log(message: String)
}

public class Satellite {
    
    public static let shared: Satellite = Satellite()
    
    public class Options {
        public var distanceFilter: CLLocationDistance = 10
        public var desiredAccuracy: CLLocationAccuracy = kCLLocationAccuracyHundredMeters
        public var cacheTime: TimeInterval = 0
        
        var cache: Bool {
            return cacheTime > 0
        }
        
        public init() {}
        
        public func clone() -> Options {
            let options = Options()
            options.cacheTime = cacheTime
            options.distanceFilter = distanceFilter
            options.desiredAccuracy = desiredAccuracy
            return options
        }
    }
    
    public var log: SatelliteLogFunc = { message in print(message) }
    public var lastLocation: CLLocation? {
        return locationManager.lastLocation
    }
    
    private var locationManager: LocationManager!
    
    public func setup(locationService: LocationService? = nil) {
        if locationService == nil {
            self.locationManager = LocationManager(locationService: SystemLocationService())
        } else {
            self.locationManager = LocationManager(locationService: locationService!)
        }
    }
    
    public func getLocationOnce(options: Options, listener: @escaping SatelliteListener) {
        self.locationManager.getLocationOnce(options: options, listener: listener)
    }
    
    public func createContinue(options: Options, listener: @escaping SatelliteListener) -> LocationHandler {
        return LocationHandler(self.locationManager, options: options, listener: listener)
    }
    
    public class LocationHandler {
        
        weak var locationManager: LocationManager?
        var options: Options
        let listener: SatelliteListener
        var holder: LocationManager.ContinueLocationHolder?
        
        init(_ locationManager: LocationManager, options: Options, listener: @escaping SatelliteListener) {
            self.locationManager = locationManager
            self.listener = listener
            self.options = options
        }
        
        public func setOptions(_ options: Options) {
            self.options = options.clone()
            if let holder = self.holder {
                self.locationManager?.updateOptions(holder: holder, options: self.options)
            }
        }
        
        public func start() {
            if self.holder != nil {
                return
            }
            self.holder = self.locationManager?.startContinue(owner: self, options: options, listener: listener)
        }
        
        public func stop() {
            if let holder = self.holder {
                self.locationManager?.stopContinue(holder: holder)
                self.holder = nil
            }
        }
        
    }
    
}
